### PR TITLE
Add MySQL storage and tests

### PR DESCRIPTION
**Pull request by lucasmrod**
Hi folks.

This PR implements a MySQL storage, issue #4.
I've added tests that can be run on any future storage (which caught a small bug in the `FileStorage`).

### CODE REVIEW

File: http/api/tokenpki.go
Comment (zwass) [inline http/api/tokenpki.go:49]: Intending to do this?

File: storage/mysql/mysql.go
Comment (zwass) [inline storage/mysql/mysql.go:162]: I don't have much context here, so please lmk if you disagree. Should this be returning an error rather than silently doing nothing? Or, is it possible that it could be intentional to set the BaseURL to an empty string?

File: storage/storagetest/storagetest.go
Comment (zwass) [inline storage/storagetest/storagetest.go:109]: Good idea. Waiting on feedback from @jessepeterson to do that?

**Review by zwass — COMMENTED** [review]
Stoked to see this up!

File: cmd/depsyncer/main.go
Comment (jessepeterson) [inline cmd/depsyncer/main.go:34]: I think this meant to be on the `flStorage` flag?

**Review by jessepeterson — COMMENTED** [review]
This is an amazing contribution! Thanks so much for this! Here's some quick thoughts and I'll try to get time soon, hopefully early next week for more. :)

Bunch of comments inline. :)

I have some reservations about having an opinionated docker-compose front and center. Let me think more on that (maybe it's as simple as placing the compose file in storage/storagetest/?).

The cmd/cli.go -> parse/storage.go is interesting. What was the thinking there?

Also - don't forget to update the docs (docs/operations-guide.md) for the new storage backend. :)

File: storage/mysql/mysql.go
Comment (jessepeterson) [reply storage/mysql/mysql.go:162]: API guard is done here: c77808e90a2965c6791a8de56a6cf61cea4c0b9b

File: http/api/tokenpki.go
Comment (lucasmrod) [reply http/api/tokenpki.go:49]: Yes. I'll remove the comment and open a separate PR for this.

File: storage/file/file.go
Comment (lucasmrod) [reply storage/file/file.go:131]: Caught by tests! :)

File: storage/file/file.go
Comment (lucasmrod) [reply storage/file/file.go:70]: A convention I've seen many projects use is to try (when possible) to always return zero-values when error is non-nil. In this case, the zero-value of a pointer is `nil`.

File: storage/mysql/mysql.go
Comment (lucasmrod) [reply storage/mysql/mysql.go:10]: Ah, good idea.

File: storage/mysql/mysql.go
Comment (lucasmrod) [reply storage/mysql/mysql.go:100]: Oh, will check this. What was the failure?

File: storage/storagetest/storagetest.go
Comment (lucasmrod) [reply storage/storagetest/storagetest.go:109]: Oh I added it but wanted to check first if we all agree it's a good idea to define it and return it in all methods when the resource (config, tokens, pki) is not found.

File: cmd/depsyncer/main.go
Comment (lucasmrod) [reply cmd/depsyncer/main.go:34]: Oops

File: storage/file/file.go
Comment (lucasmrod) [reply storage/file/file.go:24]: Convention that serves both as a "compile time check" and as a form of documentation for readers that `FileStorage` implements `storage.AllStorage` interface.

File: storage/mysql/mysql_test.go
Comment (lucasmrod) [reply storage/mysql/mysql_test.go:24]: Ah makes sense. Will give this a thought.

File: storage/mysql/mysql.go
Comment (lucasmrod) [reply storage/mysql/mysql.go:147]: I wanted to mimick `FileStorage.RetrieveConfig`:
https://github.com/micromdm/nanodep/blob/077b73dfe75f5cf5ce12601ce1ca59bbb925bb0e/storage/file/file.go#L87-L98

File: storage/mysql/mysql.go
Comment (lucasmrod) [reply storage/mysql/mysql.go:182]: Just wanted to make it clear that this nullable variable corresponds to the named return `profileUUID`.

File: storage/mysql/mysql.go
Comment (lucasmrod) [reply storage/mysql/mysql.go:150]: As per tests it looks like it's working. (with the `parseTime` option now removed)

**Comment by lucasmrod** [comment]
I believe I've address all the comments.

> I have some reservations about having an opinionated docker-compose front and center. Let me think more on that (maybe it's as simple as placing the compose file in storage/storagetest/?).

I've moved it to `storage/mysql/`. Let me know if that works.

> The cmd/cli.go -> parse/storage.go is interesting. What was the thinking there?

Originally I didn't want mysql imports in `storage/storage.go` (where we define the abstract interface), thus I moved the one parsing method to its own package `parse`.

> Also - don't forget to update the docs (docs/operations-guide.md) for the new storage backend. :)

Done.

Appreciate the feedback!

File: http/api/tokenpki.go
Comment (lucasmrod) [reply http/api/tokenpki.go]: See https://github.com/micromdm/nanodep/pull/6.

File: parse/storage.go
Comment (jessepeterson) [inline parse/storage.go:1]: I like the idea of moving this out of the `cmd/` tree. But can we rename this package to just `cli` to match https://github.com/micromdm/nanomdm/commit/ff8b7af3437525755d0d6b05d3c710590b454b6a?

File: storage/file/file.go
Comment (jessepeterson) [reply storage/file/file.go:24]: I think this creates an unnecessary coupling. How do you feel about removing it? As Go is statically typed an implementation must conform to it's use. `AllStorage` isn't particularly useful outside of `main.go` so I'd not like to couple these.

Let me know what you think. :)

File: storage/file/file.go
Comment (jessepeterson) [reply storage/file/file.go:73]: So I'm fine with this specific code change, although its a bit verbose. But in general I don't think its necessary to add a bunch of boilerplate to assure a zero-values for an error. Discussed here: https://macadmins.slack.com/archives/C07ME1284/p1663867171949419

File: storage/mysql/mysql.go
Comment (jessepeterson) [reply storage/mysql/mysql.go:172]: Hmm, good call-out. I think the file-storage should be changed to match actually:

No config (missing) = nil
Present config (including an empty BaseURL) = returned.

They're treated the same in the config retriever interface implementation, as it happens. This should probably be documented there:

https://github.com/micromdm/nanodep/blob/a7538c5b9c87bca001a5bc636de747e1b1d10adc/client/client.go#L23-L25

**Review by jessepeterson — CHANGES_REQUESTED** [review]
sorry for the unacceptably huge delay! just a few changes. would love your input on the general Go idiomatic/convention stuff — but it sounds like we might disagree on a few small things :). hopefully we can resolve!

**Comment by lucasmrod** [comment]
> Hmm, good call-out. I think the file-storage should be changed to match actually:
No config (missing) = nil
Present config (including an empty BaseURL) = returned.

I've amended the two storage implementations to behave this way, and added a test for `DefaultConfigRetreiver`.

**Comment by lucasmrod** [comment]
@jessepeterson One last change that can be added to this PR, or open a separate one if this one is merged-as is: https://github.com/fleetdm/nanodep/pull/1

**Comment by jessepeterson** [comment]
@lucasmrod let's do a separate PR and keep this contained :)

also I think you need to 'request review' or whatever - it's showing as changes still requested :)

thanks again!

**Comment by lucasmrod** [comment]
> let's do a separate PR and keep this contained :)

Cool. I'll create another PR once this one is merged to `main`.

File: client/client.go
Comment (jessepeterson) [inline client/client.go:24]: ❤️

**Review by jessepeterson — APPROVED** [review]
thanks much for this contribution @lucasmrod !

**Pull request merged**
